Expand Static Constructor C# documentation

EugeneKramer opened the pull request:
## Summary

Expands documentation on C# static constructor.

Added:
- A static constructor will run before an instance constructor
- Static field initializer will run before static constructor
- Presence of implicit static constructor
- Note of System.TypeInitializationException exception for unhandled exceptions in static class
- Is only meant to be called by the common language runtime (CLR).
- Added link to CA2121: Static constructors should be private
- Note that presence of an explicit static constructor should be documented.
-  C# language spec link for static constructor
- Constructor Design link
- Cannot be inherited/overloaded. Only appears once.
- System.Reflection.TypeAttributes.BeforeFieldInit
- Static Readonly coverage
- <s>RuntimeHelpers.RunClassConstructor - Explicit initialization method</s>

Formatted:
- Moved usage example to separate section. Usage is not a property.

Fixes https://github.com/dotnet/docs/issues/12639

EugeneKramer committed:
Expand Static Constructor C#

Added:
- A static constructor will run before an instance constructor
- Presence of implicit static constructor
- Note of System.TypeInitializationException exception for unhandled exceptions in static class
- is only meant to be called by the common language runtime (CLR).
- Added link to CA2121: Static constructors should be private
- Note that presence of an explicit static constructor should be documented.
Removed: 
- The user has no control on when the static constructor is executed in the program...it's called when the class is first accessed/referenced/instantiated.

Formatted:
- Moved usage example to separate section. Usage is not a property.

Address: https://github.com/dotnet/docs/issues/12639

EugeneKramer committed:
Add C# language spec link for static constructor

Add C# language spec link for static constructor
 [Static constructors](~/_csharplang/spec/classes.md#static-constructors)

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/static-constructors.md:
```diff
@@ -16,22 +16,30 @@ A static constructor is used to initialize any [static](../../../csharp/language
+ [!code-csharp[csProgGuideObjects#15](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideObjects/CS/Objects.cs#15)]
+ 
+## C# language specification  
+For more information, see the [Static constructors](~/_csharplang/spec/classes.md#static-constructors) section of the [C# language specification](../../language-reference/language-specification/index.md).
```
There is no *language-specification* folder in this repo anymore and the referenced index page is redirected to the one in this suggestion from the csharplang repo

```suggestion
For more information, see the [Static constructors](~/_csharplang/spec/classes.md#static-constructors) section of the [C# language specification](~/_csharplang/spec/introduction.md).
```

EugeneKramer committed:
Update docs/csharp/programming-guide/classes-and-structs/static-constructors.md

Correct C# language spec root link.

Co-Authored-By: Petr Kulikov <petr.kulikov@gmail.com>

EugeneKramer committed:
Expanded Remarks and use case

Add:
- Cannot be inherited/overloaded and there is only one.
- Static field variable initializers executed first 
- Link to constructor design guidelines
 - Example from C# lang spec - Static constructors are also a convenient place to enforce run-time checks on the type parameter that cannot be checked at compile-time via constraints
- System.Reflection.TypeAttributes.BeforeFieldInit flag not added (impacts performance)

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/static-constructors.md:
```diff
@@ -11,27 +11,46 @@ ms.assetid: 151ec95e-3c4d-4ed7-885d-95b7a3be2e7d
-- The user has no control on when the static constructor is executed in the program.  
+- A static constructor is called automatically to initialize the [class](../../../csharp/language-reference/keywords/class.md) before the first instance is created or any static members are referenced. A static constructor will run before an instance constructor. Note that a type's static constructor is called when a static method assigned to an event or a delegate is invoked and not when it is a ...
+
+- If you don't provide a static constructor to initialize static fields, the C# compiler will supply a default implicit static constructor that initializes static fields to their default value as listed in the [Default Values Table](../../../csharp/language-reference/keywords/default-values-table.md). 
```
I'd like to rephrase this. The end result (initializing values to their default) is correct, but the compiler is not required to supply a static constructor. See the [last paragraph on default values in the C# language spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/variables#default-values).

```suggestion
- If you don't provide a static constructor to initialize static fields, all static fields are initialized to their default value as listed in the [Default Values Table](../../../csharp/language-reference/keywords/default-values-table.md). 
```

EugeneKramer replied:
Appreciate the input.

That line was incorporated from [Constructors (C# Programming Guide)](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/constructors#static-constructors) :
>...If you don't provide a static constructor to initialize static fields, the C# compiler will supply a default static constructor that initializes static fields to their default value as listed in the Default Values Table.

Also, similar phrasing is used in [Static constructors and the TypeInitializationException](https://docs.microsoft.com/en-us/dotnet/api/system.typeinitializationexception?view=netframework-4.8#static-constructors-and-the-typeinitializationexception):
> ...Static constructors can be explicitly defined by a developer. If a static constructor is not explicitly defined, compilers automatically create one to initialize any static (in C#) or Shared (in Visual Basic) members of the type.

Should the text on these other pages be revised?

BillWagner replied:
😢 Yes, those pages should be updated as well. We're moving toward ensuring that the articles in these sections use language that's consistent with the spec (although often more informal). These should agree with the spec, and not imply different behavior.

Do you want to make those additional changes, or should I create an issue to update those docs and reference this conversation?

EugeneKramer replied:
I'm not sure how to best rephrase the text, so it'd be better if you were to file the issue.

Also, having looked at the IL output on [SharpLab](https://sharplab.io), I'm not sure the phrasing is incorrect.  When a static field is initialized at declaration, a `.cctor` is generated within which the static fields are initialized.

```csharp
//no .cctor
public class A{
    public static int x;
}
```
```csharp
//has .cctor 
//initializiation takes place in .cctor
public class B{
    public static int x=1;
}
```
```csharp
//has .cctor 
//initializiation takes place in .cctor
//x initialized before  y
public class C{
    public static int x=1;
    static C(){int y=2;}
}
```

BillWagner reviewed (changes requested):
Thanks for making these updates @EugeneKramer These are great additions that provide needed information about defining static constructors.

I have one suggestion for an update where the phrasing could be mis-interpreted. Let me know what you think of the suggestion, and then this is ready to :shipit:

EugeneKramer committed:
Update docs/csharp/programming-guide/classes-and-structs/static-constructors.md

Co-Authored-By: Bill Wagner <wiwagn@microsoft.com>

EugeneKramer committed:
Add static readonly and explicit initialization

Added:
- static readonly
- explicit initialization
- BeforeFieldInit purpose

EugeneKramer commented:
I've added a note for static readonly initialization and using RuntimeHelpers.RunClassConstructor for explicitly calling the static constructor.

If they're ok, feel free to :shipit:

EugeneKramer committed:
Fix phrasing

Adjust phrasing.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/static-constructors.md:
```diff
@@ -11,27 +11,55 @@ ms.assetid: 151ec95e-3c4d-4ed7-885d-95b7a3be2e7d
+
+### Explicit initialization
+
+> [!CAUTION]
```
I think that leaving this section out would be better. Readers that already know about `RunClassConstructor` would (hopefully) know they should not call the method. Other readers can remain blissfully unaware of its existence.

What were your thoughts regarding adding this?

EugeneKramer replied:
The `RunClassConstructor` was mainly added for completion's sake . It has been removed as its presence would promote misuse.

It may be worthwhile to include an example that an empty static void method would serve the same purpose.

```csharp
using System; 

public class B 
{      
    public static void TriggerStaticConstructor() {} 
	
    static B(){
        Console.WriteLine("Static constructor invoked.");
    }
} 
public class A 
{
    public static void Main(string[] args){
        B.TriggerStaticConstructor();
    }
} 
```

BillWagner replied:
Thanks for the update. I think this addition isn't necessary, so I'll :shipit: now.

EugeneKramer committed:
Removed reference to RunClassConstructor 

Removed reference to RunClassConstructor per
https://github.com/dotnet/docs/pull/12733#discussion_r292503258
